lib/pbench/server/database/models/datasets.py

```diff
@@ -857,21 +846,44 @@ class Metadata(Database.Base):
     # {"dashboard.seen": True}
     DASHBOARD = "dashboard.*"
 
-    # USER is arbitrary data saved on behalf of the owning user, as a JSON
-    # document.
+    # USER_FAVORITE allows direct access (e.g., in /datasets/list) to the
+    # user-specific "favorite" metadata. NOTE: for now, this is hardcoded as
+    # False until we have a framework for user-specific metadata.
     #
-    # Note that the hierarchical key management in the getvalue and setvalue
-    # static methods (used consistently by the API layer) allow client code
-    # to interact with these either as a complete JSON document ("user") or
-    # as a full dotted key path ("user.contact.name.first") or at any JSON
-    # layer between.
+    # {"user.favorite": False}
+    USER_FAVORITE = "user.favorite"
```

Comment:
This is cross-cutting: each user can "favorite" a distinct set of datasets. This returns the specific user-specific favorite setting for this dataset instead of requiring a separate user metadata call (once we have such a thing). I thought that was described in the comment. 😦

Reply:
I'm grasping for a framework in which to understand this.  I get the idea that the Dashboard wants an easy way to fetch the list of "favorite" datasets.  And, I totally understand that there's a matrix problem here -- "user favorite dataset" lies at the intersection of "dataset" and "user".  But, I don't think that "dataset X is a favorite of user Y" should be an attribute of a given dataset (although, structuring it that way would likely make it easy to pull the set of results out of Elasticsearch), because that requires us to modify the records when the user changes his/her mind.

I think that leaves us storing the favorite-ness outside Elasticsearch...which I think means that there is no advantage to attaching it to the dataset.

My instinct is that the list of favorite datasets should be an attribute of the user.  The Dashboard can fetch it and then fetch the individual datasets.  In real terms, the number of favorites is unlikely to be large, and, if we need to, we can provide a mechanism which allows the server to return multiple datasets (e.g., via a list of IDs) in a single query.  (Or, we can create a special request which is implicitly driven by the favorite list, but I don't like that because it probably requires tight coupling and/or architecture.)

The advantage of associating it with the user is that adding or removing users or individual favorites does not affect datasets (or require touching them in any way).  There is the issue of what happens when a user-favorited dataset (e.g., one owned by someone else) is deleted, but I presume that we can handle that when the request is made to fetch the dataset (rather than trying to update all users' metadata when the dataset is removed).

Making it user metadata also means that it can be implemented on the client side, without requiring mapping the key or otherwise special handling in the `user` namespace.

Reply:
Elasticsearch has nothing to do with this.

We have a piece of data that's both user- and dataset-specific; the future user metadata infrastructure will provide a way to get all dataset favorites through the user profile, but this provides a way to get the authenticated user's setting for a specific dataset. This provides an easy and straightforward way for the dashboard to get the information where it matters rather than doing a separate user query and then threading the two lists together.

Reply:
I guess I don't understand what the specific requirements or intended design problem is.  There are two which leap to mind immediately (and I suspect that, at different times, the Dashboard will want to do each of these things):

1. support the Dashboard marking specific results within a larger display to show which ones have been marked as the user's favorites
1. support the Dashboard fetching/displaying a list of results where only the user's favorites are listed

For the first, we could have the `/datasets/list` "decorate" the returned values to indicate which are favorited.  I don't know that that is any easier to do on the server side than it would be to do on the client side (once the client has the list of favorites).  Either way, this seems like a relatively straightforward problem to solve, without adding to the metadata on the dataset side.

The second seems like a more challenging problem.  I'm not immediately coming up with a solution other than fetching the list of favorites from the user and then fetching those results.  If we maintained as part of the dataset the list of users who had favorited it, obviously it would be pretty easy, but I don't think we want to burden individual datasets with an unbounded list of the users which had marked it as their favorite.  Of course, if the client fetched the available datasets, it could filter out any which weren't on the user's list, but that's a pretty unsatisfying solution.